### PR TITLE
Always use WASAPI's callback mode

### PR DESCRIPTION
**Pull request by MasonRemaley**
# TLDR

This PR modifies the WASAPI backend to use event based callbacks, fixing a problem where small buffers would unnecessarily underflow due to issues with how Windows handles timer resolution.

# Full Explanation

Presently, the WASAPI backend when running in shared mode defaults to a 4 second buffer, and uses a sleep condition variable to time the write call back, the intention being that you don't have to use the whole buffer.

Unfortunately, if you only use ~23-40ms of the buffer, you'll get underflows because sleep condition variables are subject to Windows' timer interval. If you use [timeBeginPeriod](https://docs.microsoft.com/en-us/windows/win32/api/timeapi/nf-timeapi-timebeginperiod) to lower the interval, or if someone else has already lowered it ([it's kind of global ish maybe](https://randomascii.wordpress.com/2020/10/04/windows-timer-resolution-the-great-rule-change/)), the problem goes away. WASAPI appears to lower the interval itself automatically, but not enough, even when modifying libsoundio to request a smaller buffer. To check the current timer interval, see [ClockRes](https://docs.microsoft.com/en-us/sysinternals/downloads/clockres).

This PR solves this problem by modifying the WASAPI backend to use `AUDCLNT_STREAMFLAGS_EVENTCALLBACK` mode. The code changes are minimal since this mode was already used for raw devices. In this mode, `WaitForSingleObject` is used to time the callbacks instead of `SleepConditionVariableCS`. I'm under the impression that `WaitForSingleObject` doesn't rely on the timer period unless it times out, but regardless, WASAPI does a better job of setting the timer period low enough for you in this mode.

With these changes, I experience no underflows at the lowest interval Windows is willing to give me (23.49ms.)

# WIP

## Clearing The Buffer

> "On systems that support clearing the buffer, this defaults to a large latency, potentially upwards of 2 seconds, with the understanding that you will call soundio_outstream_clear_buffer when you want to reduce the latency to 0."

*libsoundio docs*

There's nothing stopping me from continuing to support clearing the outstream in shared mode, and adding support for clearing it in exclusive mode, despite neither defaulting to a large buffer size with these changes.

I could add support for this, and modify the docs accordingly, or I could remove support and leave the docs alone.

It's not clear to me which is better because I don't understand the reason for defaulting to a large buffer on some backends to begin with--is there a use case where it's beneficial to give the user a significantly larger buffer than they asked for?

If you want, we can punt on this--I can do the safe thing by implementing this according to the current docs, and then we can revisit this later, I'm planning on suggesting a change to this API anyway and can bring it up again then.

## Min Frame Counts

I left the min frame count calculation alone. I noticed that for exclusive mode, the min frame count and max frame count are the same--I'm not sure if I should do the same thing here. Presumably there's no reason to not fill the entire buffer with this change in place?

### CODE REVIEW

File: src/wasapi.c [hunk 1]
```diff
@@ -1281,9 +1281,10 @@ static int outstream_do_open(struct SoundIoPrivate *si, struct SoundIoOutStreamP
         CoTaskMemFree(mix_format);
         mix_format = NULL;
         flags = osw->need_resample ? AUDCLNT_STREAMFLAGS_AUTOCONVERTPCM | AUDCLNT_STREAMFLAGS_SRC_DEFAULT_QUALITY : 0;
+        flags |= AUDCLNT_STREAMFLAGS_EVENTCALLBACK;
```
Comment (MasonRemaley) [inline src/wasapi.c:1284]: All the changes to this function are just switching up the arguments used when creating the outstream so that it uses the callback mode in shared mode instead of just exclusive

File: src/wasapi.c [hunk 2]
```diff
@@ -1408,34 +1408,22 @@ static void outstream_shared_run(struct SoundIoOutStreamPrivate *os) {
         outstream->error_callback(outstream, SoundIoErrorStreaming);
         return;
     }
-    osw->writable_frame_count = osw->buffer_frame_count - frames_used;
-    if (osw->writable_frame_count <= 0) {
+    int writable_frame_count = osw->buffer_frame_count - frames_used;
```
Comment (MasonRemaley) [inline src/wasapi.c:1411]: outstream_shared_run now works very similarly to outstream_raw_run

File: src/wasapi.c [hunk 3]
```diff
@@ -1598,12 +1586,10 @@ static int outstream_open_wasapi(struct SoundIoPrivate *si, struct SoundIoOutStr
         return SoundIoErrorNoMem;
     }
 
-    if (osw->is_raw) {
-        osw->h_event = CreateEvent(NULL, FALSE, FALSE, NULL);
-        if (!osw->h_event) {
-            outstream_destroy_wasapi(si, os);
-            return SoundIoErrorOpeningDevice;
-        }
+    osw->h_event = CreateEvent(NULL, FALSE, FALSE, NULL);
```
Comment (MasonRemaley) [inline src/wasapi.c:1589]: The rest of the changes are just using the event for signaling in both modes instead of just in raw
